mne/annotations.py

```diff
@@ -517,3 +517,121 @@ def _ensure_annotation_object(obj):
     if not isinstance(obj, Annotations):
         raise ValueError('Annotations must be an instance of '
                          'mne.Annotations. Got %s.' % obj)
+
+
+def _counter_factory():
+    """create a callable counter that takes 1 parameter.
+
+    It can be used as follows::
+
+        >>> my_counter = _counter_factory()
+        >>> my_counter(None)
+        1
+        >>> my_counter('usless string')
+        2
+        >>> my_counter(42)
+        3
+        >>> [my_counter(x) for x in ['a', None, 42]]
+        [4, 5, 6]
+        >>> my_other_counter = _counter_factory()
+        >>> [my_other_counter(x) for x in ['a', None, 42]]
+        [1, 2, 3]
+
+    """
+    class Counter():
+        count = 1
+
+        def __call__(self, *args, **kargs):
+            c = self.count
+            self.count += 1
+            return c
+
+    return Counter()
+
+
+@verbose
+def events_from_annotations(raw, event_id=None, regexp=None,
+                            event_id_func=None, verbose=None):
+    """Get events and event_id from an Annotations object.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw data for which Annotations are defined.
+
+    event_id : dict | None
+        Dictionary of string keys and integer values as used in mne.Epochs
+        to map annotation descriptions to integer event codes. Only the
+        keys present will be mapped and the annotations with other descriptions
+        will be ignored. If None, all descriptions of annotations are mapped
+        and assigned arbitrary unique integer values.
+
+    regexp : str | None
+        Regular expression used to filter the annotations whose
+        descriptions is a match.
+
+    event_id_func : None | callable
+        What to do for events not found in ``event_id``. Must take one
+        argument and return an ``int``. If None, ``event_id_func`` behaves like
+        a counter that increments every time is called.
+
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see
+        :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>`
+        for more). Defaults to self.verbose.
+
+    Returns
+    -------
+    events : ndarray, shape (n_events, 3)
+        The events.
+    event_id : dict
+        The event_id variable that can be passed to Epochs.
+    """
+    import re
+
+    if raw.annotations is None:
+        return np.empty((0, 3), dtype=int), event_id
+
+    annotations = raw.annotations
+
+    inds = raw.time_as_index(annotations.onset, use_rounding=True)
```

Comment:
I doubt this works as `raw.time_as_index` does not take into account first_samp while `annotations.onset` does

Reply:
It does since onset is taken from meas_date

Reply:
but time_as_index ignores meas_date and assumes it's always None. Yes I know it's bad
but it's as it is right now unless I overlooked it.

Reply:
Maybe #5523 illustrates the problem

Reply:
what about `inds = raw.time_as_index(annotations.onset, use_rounding=True, origin='meas_date')`